### ISSUE TITLE
Since 04.05.2019 Jenkins Dependency Check (4.0.2) is throwing "IllegalArgumentException: Comparison method violates its general contract!"

### ISSUE THREAD

**Issue by kaszuster** [dependency-check/DependencyCheck#1900]
**Describe the bug**
Since 04.05.2019 we can not run the Jenkins Dependency Check Plugin Version 4.0.2 anymore.

To me it looks like that the Updated Database (Dependency-Check NVD update) is not compatible with Version 4.0.2. Creating a demo project (containing vulnerable dependencies) and running Dependency Check Version 5.0.0-M2 (as Maven plugin) runs as expected, switch the Version Back to 4.0.2 fails with the same error that Jenkins has.

** Version of dependency-check used **
4.0.2 - Jenkins Plugin (original test) => Error
4.0.2 - Maven Plugin => Error
5.0.0-M2 - Maven Plugin => Success

**Log file**
```
...
java.lang.IllegalArgumentException: Comparison method violates its general contract!
	at java.util.ComparableTimSort.mergeLo(ComparableTimSort.java:744)
	at java.util.ComparableTimSort.mergeAt(ComparableTimSort.java:481)
	at java.util.ComparableTimSort.mergeCollapse(ComparableTimSort.java:406)
	at java.util.ComparableTimSort.sort(ComparableTimSort.java:213)
	at java.util.Arrays.sort(Arrays.java:1312)
	at java.util.Arrays.sort(Arrays.java:1506)
	at java.util.ArrayList.sort(ArrayList.java:1462)
	at java.util.Collections.sort(Collections.java:141)
	at org.owasp.dependencycheck.dependency.Vulnerability.getVulnerableSoftware(Vulnerability.java:264)
	at sun.reflect.GeneratedMethodAccessor605.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.velocity.util.introspection.UberspectImpl$VelMethodImpl.doInvoke(UberspectImpl.java:395)
	at org.apache.velocity.util.introspection.UberspectImpl$VelMethodImpl.invoke(UberspectImpl.java:384)
	at org.apache.velocity.runtime.parser.node.ASTMethod.execute(ASTMethod.java:173)
Caused: org.apache.velocity.exception.MethodInvocationException: Invocation of method 'getVulnerableSoftware' in  class org.owasp.dependencycheck.dependency.Vulnerability threw exception java.lang.IllegalArgumentException: Comparison method violates its general contract! at templates/htmlReport.vsl[line 902, column 23]
	at org.apache.velocity.runtime.parser.node.ASTMethod.handleInvocationException(ASTMethod.java:243)
	at org.apache.velocity.runtime.parser.node.ASTMethod.execute(ASTMethod.java:187)
	at org.apache.velocity.runtime.parser.node.ASTReference.execute(ASTReference.java:280)
	at org.apache.velocity.runtime.parser.node.ASTReference.value(ASTReference.java:567)
	at org.apache.velocity.runtime.directive.Foreach.render(Foreach.java:318)
	at org.apache.velocity.runtime.parser.node.ASTDirective.render(ASTDirective.java:207)
	at org.apache.velocity.runtime.parser.node.ASTBlock.render(ASTBlock.java:72)
	at org.apache.velocity.runtime.parser.node.SimpleNode.render(SimpleNode.java:342)
	at org.apache.velocity.runtime.parser.node.ASTIfStatement.render(ASTIfStatement.java:106)
	at org.apache.velocity.runtime.parser.node.ASTBlock.render(ASTBlock.java:72)
	at org.apache.velocity.runtime.parser.node.ASTIfStatement.render(ASTIfStatement.java:87)
	at org.apache.velocity.runtime.parser.node.ASTBlock.render(ASTBlock.java:72)
	at org.apache.velocity.runtime.directive.Foreach.render(Foreach.java:420)
	at org.apache.velocity.runtime.parser.node.ASTDirective.render(ASTDirective.java:207)
	at org.apache.velocity.runtime.parser.node.ASTBlock.render(ASTBlock.java:72)
	at org.apache.velocity.runtime.parser.node.ASTIfStatement.render(ASTIfStatement.java:87)
	at org.apache.velocity.runtime.parser.node.ASTBlock.render(ASTBlock.java:72)
	at org.apache.velocity.runtime.directive.Foreach.render(Foreach.java:420)
	at org.apache.velocity.runtime.parser.node.ASTDirective.render(ASTDirective.java:207)
	at org.apache.velocity.runtime.parser.node.SimpleNode.render(SimpleNode.java:342)
	at org.apache.velocity.runtime.RuntimeInstance.render(RuntimeInstance.java:1378)
	at org.apache.velocity.runtime.RuntimeInstance.evaluate(RuntimeInstance.java:1314)
	at org.apache.velocity.app.VelocityEngine.evaluate(VelocityEngine.java:272)
	at org.owasp.dependencycheck.reporting.ReportGenerator.processTemplate(ReportGenerator.java:355)
	at org.owasp.dependencycheck.reporting.ReportGenerator.processTemplate(ReportGenerator.java:313)
	at org.owasp.dependencycheck.reporting.ReportGenerator.write(ReportGenerator.java:258)
	at org.owasp.dependencycheck.reporting.ReportGenerator.write(ReportGenerator.java:229)
	at org.owasp.dependencycheck.Engine.writeReports(Engine.java:1118)
	at org.owasp.dependencycheck.Engine.writeReports(Engine.java:1135)
	at org.jenkinsci.plugins.DependencyCheck.DependencyCheckExecutor.generateExternalReports(DependencyCheckExecutor.java:188)
	at org.jenkinsci.plugins.DependencyCheck.DependencyCheckExecutor.call(DependencyCheckExecutor.java:107)
	at org.jenkinsci.plugins.DependencyCheck.DependencyCheckExecutor.call(DependencyCheckExecutor.java:46)
	at hudson.remoting.LocalChannel.call(LocalChannel.java:45)
	at org.jenkinsci.plugins.DependencyCheck.AbstractDependencyCheckBuilder.perform(AbstractDependencyCheckBuilder.java:85)
	at org.jenkinsci.plugins.DependencyCheck.DependencyCheckBuilder.perform(DependencyCheckBuilder.java:206)
	at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:80)
	at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:67)
	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Finished: FAILURE
```

**To Reproduce**
- Create Java Project (open jdk 8)
- Add vulnerable dependency e.g. `org.im4java:im4java:1.4.0` 
- Add Maven Plugin:
```
<plugin>
    <groupId>org.owasp</groupId>
    <artifactId>dependency-check-maven</artifactId>
    <version>4.0.2</version>
    <executions>
        <execution>
            <goals>
                <goal>check</goal>
            </goals>
        </execution>
    </executions>
</plugin>
```
- Run `mvn clean install`

**Expected behavior**
Do not throw `java.lang.IllegalArgumentException: Comparison method violates its general contract!`

**Additional context**
I have read that Version 5.0.0 will be released soon and will contain breaking changes, maybe this is one of these changes? For now i will deactivate this buildstep in my pipeline and wait for the release of V5.

**Comment by jeremylong** [comment]
This has been resolved in the 5.x snapshots and will be included in the 5.0.0 release.

**Comment by isqo** [comment]
you can get around this bug by disabling html report generation if you don't need it. (includeHtmlReports: false)
